cirq-core/cirq/optimizers/two_qubit_to_sqrt_iswap.py

```diff
@@ -0,0 +1,413 @@
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    x, y, z = kak.interaction_coefficients
+    a0, a1 = kak.single_qubit_operations_before
+    c0, c1 = kak.single_qubit_operations_after
+
+    # Computed gate parameters: Eq. 4, 6, 7, 8 of the paper
+    # range limits added for robustness to numerical error
+    def safe_arccos(v):
+        return np.arccos(np.clip(v, -1, 1))
+
+    def nonzero_sign(v):
+        return -1 if v < 0 else 1
+
+    _c = np.clip(
+        np.sin(x + y - z) * np.sin(x - y + z) * np.sin(-x - y - z) * np.sin(-x + y + z), 0, 1
+    )
+    alpha = safe_arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) + 2 * np.sqrt(_c))
+    beta = safe_arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) - 2 * np.sqrt(_c))
+    _4ccs = 4 * (np.cos(x) * np.cos(z) * np.sin(y)) ** 2  # Intermediate value
+    gamma = safe_arccos(
+        nonzero_sign(z)
+        * np.sqrt(_4ccs / (_4ccs + np.clip(np.cos(2 * x) * np.cos(2 * y) * np.cos(2 * z), 0, 1)))
+    )
+
+    # Inner single-qubit gates: Fig. 4 of the paper
+    b0 = (
+        protocols.unitary(ops.rz(-gamma))
+        @ protocols.unitary(ops.rx(-alpha))
+        @ protocols.unitary(ops.rz(-gamma))
+    )
+    b1 = protocols.unitary(ops.rx(-beta))
+
+    # Compute KAK on the decomposition to determine outer single-qubit gates
+    # There is no known closed form solution for these gates
+    u_sqrt_iswap = protocols.unitary(ops.SQRT_ISWAP)
+    u = u_sqrt_iswap @ np.kron(b0, b1) @ u_sqrt_iswap  # Unitary of decomposition
+    kak_fix = linalg.kak_decomposition(u, atol=atol / 10, rtol=0, check_preconditions=False)
+    d0, d1 = kak_fix.single_qubit_operations_before
+    e0, e1 = kak_fix.single_qubit_operations_after
+
+    return [  # Pairs of single-qubit unitaries, SQRT_ISWAP between each is implied
+        (d0.T.conj() @ a0, d1.T.conj() @ a1),
+        (b0, b1),
+        (c0 @ e0.T.conj(), c1 @ e1.T.conj()),
+    ], kak.global_phase / kak_fix.global_phase
+
+
+def _decomp_3sqrt_iswap_matrices(
+    kak: 'cirq.KakDecomposition',
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Returns the single-qubit matrices for the 3-SQRT_ISWAP decomposition.
+
+    Assumes any canonical x, y, z.  Three sqrt-iSWAP gates are only needed if
+    x < y + |z|.  Only two are needed for other gates (most cases).
+
+    References:
+        Towards ultra-high fidelity quantum operations: SQiSW gate as a native
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    x, y, z = kak.interaction_coefficients
+    f0, f1 = kak.single_qubit_operations_before
+    g0, g1 = kak.single_qubit_operations_after
+
+    # Find x1, y1, z1, x2, y2, z2
+    # such that x1+x2=x, y1+y2=y, z1+z2=z
+    # where x1, y1, z1 are implementable by one sqrt-iSWAP gate
+    # and x2, y2, z2 implementable by two sqrt-iSWAP gates
+    # No error tolerance needed
+    ieq1 = y > np.pi / 8
+    ieq2 = z < 0
+    if ieq1:
```

Comment:
I'll do that.  I couldn't get this part of the algorithm from the paper to work so this is actually an improvement over the paper that supports decomposing any gate into 3-sqrt-iSWAP.

Reply:
To clarify your statement, this doesn't "recompute" kak, it just uses `kak_canonicalize_vector` to do the canonicalization automatically instead of manually like in Algorithm 1.

Reply:
Comment added.